schemas: specify Any and AnyScalar

rvagg opened the pull request:
First attempt at specifying `Any`, also introduces `AnyScalar` and adds some words about `&` usage into `TypeLink` comments. The example.ipldsch is also adjusted to include links and it has the missing `"expectedType"` where it's `Any` in the first instance.

vmx commented:
Sounds good to me, but it should really be approved by @warpfork.

warpfork commented on schemas/schema-schema.ipldsch:
```diff
@@ -135,6 +135,35 @@ type RepresentationKind enum {
+## Links are left out of this union for now, and therefore any element in
+## the schema DSL (`&Any`) that uses it. The placement of links should be
+## intentional when designing schemas.
+##
```
```suggestion
+## Note that "Any" also shows up as a TypeName string literal in some parts
+## of the schema-schema, such as being the implicit value of the TypeName
+## for the expectedType field of a TypeLink.
+##
```

warpfork commented on schemas/schema-schema.ipldsch:
```diff
@@ -135,6 +135,35 @@ type RepresentationKind enum {
+	| TypeFloat "float"
+	| TypeMap "map"
+	| TypeList "list"
+} representation kinded
```
Here and below, wee typo: kinded unions need no quotes, because kinds are keywords.  (The quoted strings wouldn't show up anywhere in serial form, either, so quoting them would be confusing.)

warpfork replied:
Also, wrong level of meta.   'Any' contains any *value*, not the description of any *type*.  Thus:

```suggestion
-       | TypeBool "bool"
-       | TypeString "string"
-       | TypeBytes "bytes"
-       | TypeInt "int"
-       | TypeFloat "float"
-       | TypeMap "map"
-       | TypeList "list"
+       | Bool bool
+       | String string
+       | Bytes bytes
+       | Int int
+       | Float float
+       | Map map
+       | List list
```

warpfork replied:
Oh shoot.  There's no great definition of `Map` for this purpose.  It would have to be something like `{String:Any}` but that seems a bit shakey itself for putting into the prelude of known type names.  I'm not sure what to do about this.

warpfork commented:
@rvagg, I think you'd suggested in earlier discussion that you're suspicious "Any" is more complicated than this, and might need a special construct, and I'm thinking you were right.

A link of '&Any' can certainly point to a union or a struct.  This doesn't fit with the idea of a kinded union.  We'd seem to need a special construct for that.

And that does seem to be what the use in e.g. HAMTs internals requires.  Link targets in general, too.

`AnyScalar`, on the other hand, seems to still be clear without this (nor run into any of the other tricky issues).  So maybe we should proceed with just that for now -- and it is, fortunately, the only thing we really want in the schema-schema itself.

rvagg commented:
I've removed `Any` but left `AnyScalar`.

I've also changed `Link`'s `expectedType` to `String` instead of `TypeTerm` because we can't supply an implicit for a non-scalar. Parsers are just going to have to special-case that one. schema-schema.ipldsch now parsers properly and the .json is updated too here (it didn't parse properly with an implicit for `TypeTerm`).

@warpfork see lastest changes.

warpfork commented:
Ah right that's the other thing I was trying to remember to mention.

We can make `Link`'s `expectedType` a `TypeName` rather than a `TypeTerm`, and that clears that up because then the implicit is just scalar content again.  I was a doofus in saying `TypeTerm` at earlier points -- forgot the difference between the two.

rvagg commented:
> We can make Link's expectedType a TypeName rather than a TypeTerm

That still introduces complications. The parser has to see the type that the "implicit" needs to be coerced to and make sure it can perform such coercion. If we restrict "implicit" usage to only AnyScalar types then it's straightforward. If we want to make TypeName able to have an implicit then we need to either have TypeName declared before this block (first introduction of ordering, first declare-before-use) or do a double pass through the parsed "AST" to do the type checking for implicits after we've loaded all the types and can determine that TypeName is really just a string kind and we can coerce that.

So ... How about we just a restriction for now that you can only use "implicit" on AnyScalar typed struct properties and let Link's expectedType be a string to satisfy that?

warpfork commented:
Hm.  I see the concern about a double pass, but think it might be fine and justified -- I'm pretty sure there's other parts of schema validation that *only* are possible in a second pass anyway, so we'll already have one (e.g. the reference checks for type names already poke the exact same nerve, and declare-before-use isn't an option at all since that case is freely recursive).

But I'm also fine with pushing that off down the road if you'd rather.  You're closer to the implementation of that than I am at the moment, I think.

rvagg commented:
> But I'm also fine with pushing that off down the road if you'd rather

:thumbsup: this change doesn't preclude further iteration in that direction.
